File: Prototype/Exemplos/Ex1_Prototype/Ex1.java
/**
 * PADRÃO: PROTOTYPE
 * 
 * O padrão Prototype permite copiar objetos existentes sem tornar seu código dependente de suas classes.
 * Ele define uma interface 'clone' que deve ser implementada pelas classes que permitem serem duplicadas.
 */

// 1. INTERFACE PROTOTYPE
interface Prototype {
    Prototype clone();
}

// 2. PROTÓTIPO CONCRETO
class Pessoa implements Prototype {
    private int id;
    private String nome;
    private int idade;

    public Pessoa(int id, String nome, int idade) {
        this.id = id;
        this.nome = nome;
        this.idade = idade;
    }

    // Método de clonagem
    @Override
    public Pessoa clone() {
        return new Pessoa(this.id, this.nome, this.idade);
    }

    public void setNome(String nome) { this.nome = nome; }

    @Override
    public String toString() {
        return "Pessoa [id=" + id + ", nome=" + nome + ", idade=" + idade + "]";
    }
}

// 3. REGISTRY (Gerenciador de Protótipos)
class GerenciaPessoa {
    private java.util.Map<Integer, Pessoa> pessoas = new java.util.HashMap<>();

    public void addPessoa(int id, String nome, int idade) {
        pessoas.put(id, new Pessoa(id, nome, idade));
    }

    public Pessoa getPessoaById(int id) {
        Pessoa p = pessoas.get(id);
        return (p != null) ? p.clone() : null;
    }

    public Pessoa getOriginal(int id) {
        return pessoas.get(id);
    }
}

// 4. CLIENTE
public class Ex1 {
    public static void main(String[] args) {
        GerenciaPessoa gerencia = new GerenciaPessoa();

        gerencia.addPessoa(1, "João", 30);
        gerencia.addPessoa(2, "Maria", 25);

        // Clonando
        Pessoa pessoaClone = gerencia.getPessoaById(1);
        if (pessoaClone != null) {
            pessoaClone.setNome("Novo João");
        }

        System.out.println("Pessoa Original: " + gerencia.getOriginal(1));
        System.out.println("Pessoa Clone: " + pessoaClone);
    }
}

File: Prototype/Exemplos/Ex1_Prototype/Ex1.py
"""
PADRÃO: PROTOTYPE

O padrão Prototype permite a cópia de objetos existentes sem tornar seu código
dependente de suas classes concretas. Ele desacopla o processo de clonagem do cliente.
"""

import copy

# 1. PROTÓTIPO CONCRETO
class Pessoa:
    def __init__(self, id_pessoa, nome, idade):
        self.id = id_pessoa
        self.nome = nome
        self.idade = idade

    # Método de clonagem usando o módulo 'copy' do Python
    def clone(self):
        return copy.copy(self)

    def __str__(self):
        return f"Pessoa [id={self.id}, nome={self.nome}, idade={self.idade}]"

# 2. REGISTRY (Gerenciador de Protótipos)
class GerenciaPessoa:
    def __init__(self):
        self.pessoas = {}

    def addPessoa(self, id_pessoa, nome, idade):
        self.pessoas[id_pessoa] = Pessoa(id_pessoa, nome, idade)

    def getPessoaById(self, id_pessoa):
        pessoa_original = self.pessoas.get(id_pessoa)
        if pessoa_original:
            return pessoa_original.clone()
        return None

# 3. CLIENTE
if __name__ == "__main__":
    gerencia = GerenciaPessoa()

    gerencia.addPessoa(1, "João", 30)
    gerencia.addPessoa(2, "Maria", 25)

    # Clonando a primeira pessoa
    pessoa_clone = gerencia.getPessoaById(1)
    if pessoa_clone:
        pessoa_clone.nome = "Novo João"

    # Exibindo resultados
    print(f"Pessoa Original: {gerencia.pessoas[1]}")
    print(f"Pessoa Clone: {pessoa_clone}")
